plugins/native/src/main/java/org/qbicc/plugin/native_/StructMemberAccessBasicBlockBuilder.java

```diff
@@ -19,6 +20,9 @@ public StructMemberAccessBasicBlockBuilder(CompilationContext context, BasicBloc
 
     public ValueHandle referenceHandle(Value reference) {
         if (reference.getType() instanceof CompoundType) {
+            if (reference instanceof Deref d) {
+                d.excludeLoad();
+            }
```

Comment:
I think neither this nor the following transformation `reference.getValueHandle()` are correct: not every argument to `referenceHandle` is guaranteed to have a value handle in turn.

I'm thinking we probably want these transformations in this class (I'll also include the valid one we already have here):

* `instanceFieldOf(instance, field)` ⇒ `memberOf(instance, member)` when `instance` is has a type of `CompoundType` (this one is already implemented & valid)
* `referenceHandle(Deref(x))` ⇒ `pointerHandle(x)` (this should work in all cases because `x` must be a pointer for `Deref` to work)

Transitively, these two should automatically yield `instanceFieldOf(ReferenceHandle(Deref(x)), field)` ⇒ `memberOf(PointerHandle(x), member)` when `x` is a pointer to a structure, without having to do any additional transformations.  This should partially mitigate the problem, but not fully.

Still missing is a general solution a value is an immediate structure value: `MemberOf(ReferenceHandle(immediateStructure),member)`.  We know that we can't pass things that are not references through to the last `referenceHandle(xxx)` or else we'll get a class cast exception on the value type.  So we have to transform this one here, somehow.

Note that the same problem applies for `ElementOf(ReferenceHandle(thingThatIsNotAReference),index)` in the exact same way for arrays that `MemberOf` applies for compound types.

The ideal transformation would work like this:

* `memberOf(ReferenceHandle*(sv),member)` ⇒ `extractMember(sv,member)` where `sv` is a value of type `CompoundType` and `*` indicates an invalid/impossible node
* `elementOf(ReferenceHandle*(av),index)` ⇒ `extractElement(av,index)` where `av` is a value of type `ArrayType` and `*` indicates an invalid/impossible node

However we can't actually construct these incorrect nodes.  There are two possible solutions I can think of:

1. Add another node type - call it `ImmediateHandle` - which will throw an exception if you try to take its pointer, or lower it, or use it in any way, which acts as a replacement for `ReferenceHandle` whenever the value is an immediate non-reference/non-pointer value, and implement the transformations in terms of this element (e.g. `Load(MemberOf(ImmediateHandle(x),m))` ⇒ `ExtractMember(x,m)`; with this approach we could leak these handles into the graph though, and cause late errors to be raised which might be harder to debug
2. Modify `MethodParser` to evaluate values before passing them to `referenceHandle` - if the value is an immediate value, then use the erstwhile super-secret (i.e. unused) `ExtractField` instead for structures (and `ExtractElement` for arrays); I favor this approach because it does not rely on having an invalid intermediate graph state (note that while `MethodParser` is a huge class, there are only six sites that actually do loads, and the only ones we actually care about are the ones for `OP_GETFIELD` and the `OP_*ALOAD` opcodes)

So let's assume for discussion we go with solution 2, and modify `MethodParser` to use `ExtractField` instead of `Load(InstanceFieldOf(...))` when the instance is not a reference.  In that case we can now easily implement several more transformations:

* `extractField(x, fld)` ⇒ `extractMember(x, member)` where `x.getType()` is `CompoundType`; this would be similar to the way `instanceFieldOf` is already implemented to resolve fields to members below
* `extractMember(Deref(ptr), member)` ⇒ `deref(memberOf(pointerHandle(ptr), member)` (i.e. recursively factor `Deref` to the outermost place)
* `extractElement(Deref(ptr), index)` ⇒ `deref(elementOf(pointerHandle(ptr), index)` (same thing but for arrays - note that some array handling is in `PointerBasicBlockBuilder` for historical reasons and probably needs cleaning up there)

These transformations all together should combine to reduce the problem where we are passing around whole structures (but probably not eliminate it).  I suspect we will have to deal with `Load(var)` where `var` is a whole compound structure next (typically in a local variable).  With the aforementioned `MethodParser` changes we could try the following additional transformations:

* `extractMember(Load(x,NONE),member)` ⇒ `load(memberOf(x,member),NONE)` for any `x`, but only when atomicity mode is `NONE`
* `extractElement(Load(x,NONE),index)` ⇒ `load(elementOf(x,index),NONE)` with same restriction

These are slightly sketchy because `Load` is ordered (even though it has no atomicity constraint) and so we're bending the rules a bit here by transforming a non-ordered node into an ordered node, which could go south on us in subtle or bizarre ways later on.  So this leads me to think that loads with no atomicity mode might be better implemented as a new node `LoadPlain`, but I'm not decided on that yet; anyway that would have a high similarity to `Deref` if `Deref` operated on handles instead of pointer values (which is a change that we might want to try *after* this is all sorted out).  So this is something we'll definitely revisit one way or another after this PR is done.